nntrainer/layers/cl_layers/concat_cl.cpp

```diff
@@ -0,0 +1,540 @@
+  const TensorDim &input_dim_0 = input_dims[SINGLE_INOUT_IDX];
+  unsigned int concat_dim_val = input_dim_0.getTensorDim(concat_dimension);
+
+  for (unsigned int idx = 1; idx < input_dims.size(); ++idx) {
+    const TensorDim &dim = input_dims[idx];
+
+    for (unsigned int i = 0; i < ml::train::TensorDim::getNumDim(); ++i) {
+      if (i == concat_dimension)
+        continue;
+      NNTR_THROW_IF(input_dim_0[i] != dim[i], std::runtime_error)
+        << "Error: concat layer requires same shape from all input layers "
+           "along non-concat dimension";
+    }
+    concat_dim_val += dim[concat_dimension];
+  }
+
+  TensorDim output_dim = input_dim_0;
+  output_dim.setTensorDim(concat_dimension, concat_dim_val);
+
+  context.setOutputDimensions({output_dim});
+
+  /**
+   * Setup output_reshape_helper to which output will be reshaped in forwarding
+   * to facilitate easier processing.
+   *
+   * The helper shape consolidates all the dimensions before the axis
+   * together and all the dimensions after the axis to facilitate
+   * easier splitting of the data.
+   */
+  leading_helper_dim = 1;
+  output_reshape_helper.channel(1);
+  output_reshape_helper.height(1);
+  output_reshape_helper.width(1);
+  for (unsigned int idx = 1; idx < concat_dimension; ++idx) {
+    leading_helper_dim *= output_dim.getTensorDim(idx);
+  }
+
+  output_reshape_helper.height(output_dim.getTensorDim(concat_dimension));
+
+  for (unsigned int idx = concat_dimension + 1;
+       idx < ml::train::TensorDim::getNumDim(); ++idx) {
+    output_reshape_helper.width(output_reshape_helper.width() *
+                                output_dim.getTensorDim(idx));
+  }
+
+  /**
+   * Setup input_reshape_helper to which inputs will be reshaped in forwarding
+   * to facilitate easier processing.
+   */
+  input_reshape_helper.resize(input_dims.size());
+  for (unsigned int idx = 0; idx < input_reshape_helper.size(); idx++) {
+    input_reshape_helper[idx] = output_reshape_helper;
+    input_reshape_helper[idx].height(
+      input_dims[idx].getTensorDim(concat_dimension));
+  }
+
+  setBatch(input_dims[SINGLE_INOUT_IDX].batch());
+}
+
+void ConcatLayerCl::forwarding(RunLayerContext &context, bool training) {
+  Tensor &out = context.getOutput(SINGLE_INOUT_IDX);
+  const Tensor &in1 = context.getInput(INPUT_IDX_1);
+  const Tensor &in2 = context.getInput(INPUT_IDX_2);
+  ConcatProcess(in1, in2, out, context);
+}
+
+void ConcatLayerCl::incremental_forwarding(RunLayerContext &context,
+                                           unsigned int from, unsigned int to,
+                                           bool training) {
+  /**
+   * @todo create another kernel for incremental_forwarding taking into
+   * consideration from and to values
+   */
+}
+
+opencl::Kernel ConcatLayerCl::kernel_concat;
+opencl::Kernel ConcatLayerCl::kernel_concat_fp16;
+
+void ConcatLayerCl::ConcatProcess(Tensor const &in1, Tensor const &in2,
```

Comment:
Don't you use input_reshape_helper in forwarding process?
This code seems to applicable for `concat_dimension=3`.
As I understood, the `input_reshape helper` is constructed to simplify the forwarding process whatever its `concat_dimension` is. However, this code doesn't use it during forwarding, while it constructed the input_reshape_helper in `finalize()`. If so, it doesn't seem to coherent with calcDerivative.